File: packages/database/src/schemas/students.ts
import { relations } from 'drizzle-orm'
import { integer, jsonb, numeric, pgTable, varchar } from 'drizzle-orm/pg-core'
import { TIMESTAMP_COLUMNS } from '../utils.js'
import { academicYears } from './academic-years.js'
import { accounts } from './accounts.js'
import { assignments } from './assignments.js'
import { majors } from './majors.js'
import { openingSubjectAdditionalStudents } from './opening-subject-additional-students.js'
import { professors } from './professors.js'

/**
 * Stores student specific information.
 */
export const students = pgTable('students', {
  accountId: varchar('account_id', { length: 26 })
    .notNull()
    .primaryKey()
    .references(() => accounts.id),
  majorId: varchar('major_id', { length: 26 })
    .notNull()
    .references(() => majors.id),
  academicYearId: varchar('academic_year_id', { length: 26 })
    .notNull()
    .references(() => academicYears.id),
  professorId: varchar('professor_id', { length: 26 })
    .notNull()
    .references(() => professors.accountId),

  /**
   * A meaningful student id easy for remembering.
   */
  studentId: varchar('student_id', { length: 32 }).notNull().unique(),

  /**
   * Student's behavioral score, if this score falls below 40, a student's status
   * will be removed.
   */
  behavioralScore: integer('behavioral_score').notNull().default(100),

  /**
   * Json storing information about a student including more info needed for a student
   * status.
   */
  status: jsonb('status').notNull(),

  /**
   * Student's GPA from previous school.
   */
  previousGpa: numeric('previous_gpa', { precision: 3, scale: 2 }).notNull(),
  ...TIMESTAMP_COLUMNS
})

export const studentsRelations = relations(students, ({ many }) => ({
  openingSubjectAdditionalStudents: many(openingSubjectAdditionalStudents),
  assignments: many(assignments)
}))

export type Studying = {
  type: 'Studying'
}

export type RestingForTreatment = {
  type: 'RestingForTreatment'
  semester_id: string
}

export type Resting = {
  type: 'Resting'
  semester_id: string
}

export type ForcedToRest = {
  type: 'ForcedToRest'
  semester_id: string
}

export type RetiredByInsufficientGrades = {
  type: 'RetiredByInsufficientGrades'
  retired_at: string
}

export type RetiredByNotPaying = {
  type: 'RetiredByNotPaying'
  retired_at: string
}

export type RetiredByNotSigning = {
  type: 'RetiredByNotSigning'
  retired_at: string
}

export type RetiredBySigningButNotPaying = {
  type: 'RetiredBySigningButNotPaying'
  retired_at: string
}

export type RetiredByRunningOutOfTimeToStudy = {
  type: 'RetiredByRunningOutOfTimeToStudy'
  retired_at: string
}

export type RetiredByHavingInsufficientEnglishGrades = {
  type: 'RetiredByHavingInsufficientEnglishGrades'
  retired_at: string
}

export type Resigned = {
  type: 'Resigned'
  resigned_at: string
  reason: string
}

export type Fired = {
  type: 'Fired'
  fired_at: string
  reason: string
}

export type Deceased = {
  type: 'Deceased'
  deceased_at: string
}

export type Graduated = {
  type: 'Graduated'
  graduated_at: string
}

export type StudentStatus =
  | Studying
  | RestingForTreatment
  | Resting
  | ForcedToRest
  | RetiredByInsufficientGrades
  | RetiredByNotPaying
  | RetiredByNotSigning
  | RetiredBySigningButNotPaying
  | RetiredByRunningOutOfTimeToStudy
  | RetiredByHavingInsufficientEnglishGrades
  | Resigned
  | Fired
  | Deceased
  | Graduated

File: packages/api/src/schema/students.rs
use sea_query::Iden;
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

#[derive(Iden)]
pub enum Students {
    Table,
    AccountId,
    MajorId,
    AcademicYearId,
    ProfessorId,
    StudentId,
    BehavioralScore,
    Status,
    PreviousGpa,
    CreatedAt,
    UpdatedAt,
}

/// Struct denoting the user's type in a `jsonb` column.
#[derive(Deserialize, Serialize)]
#[serde(tag = "type")]
#[serde(rename_all = "kebab-case")]
pub enum StudentStatus {
    Studying,
    RestingForTreatment {
        semester_id: String,
    },
    Resting {
        semester_id: String,
    },
    ForcedToRest {
        semester_id: String,
    },
    RetiredByInsufficientGrades {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    RetiredByNotPaying {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    RetiredByNotSigning {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    RetiredBySigningButNotPaying {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    RetiredByRunningOutOfTimeToStudy {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    RetiredByHavingInsufficientEnglishGrades {
        #[serde(with = "time::serde::iso8601")]
        retired_at: OffsetDateTime,
    },
    Resigned {
        reason: String,
        #[serde(with = "time::serde::iso8601")]
        resigned_at: OffsetDateTime,
    },
    Fired {
        reason: String,
        #[serde(with = "time::serde::iso8601")]
        fired_at: OffsetDateTime,
    },
    Deceased {
        #[serde(with = "time::serde::iso8601")]
        deceased_at: OffsetDateTime,
    },
    Graduated {
        #[serde(with = "time::serde::iso8601")]
        graduated_at: OffsetDateTime,
    },
}
